src/toast/ops/conviqt.py

```diff
@@ -0,0 +1,693 @@
+        timer = Timer()
+        timer.start()
+        pnt = conviqt.Pointing(len(theta))
+        if pnt._nrow > 0:
+            arr = pnt.data()
+            arr[:, 0] = phi
+            arr[:, 1] = theta
+            arr[:, 2] = psi
+        if verbose:
+            timer.report_clear(f"pack input array for detector {det}")
+        return pnt
+
+    def convolve(self, sky, beam, detector, pnt, det, verbose):
+        timer = Timer()
+        timer.start()
+        convolver = conviqt.Convolver(
+            sky,
+            beam,
+            detector,
+            self.pol,
+            self.lmax,
+            self.beammmax,
+            self.order,
+            self.verbosity,
+            self.comm,
+        )
+        convolver.convolve(pnt)
+        if verbose:
+            timer.report_clear(f"convolve detector {det}")
+
+        # The pointer to the data will have changed during
+        # the convolution call ...
+
+        if pnt._nrow > 0:
+            arr = pnt.data()
+            convolved_data = arr[:, 3].astype(np.float64)
+        else:
+            convolved_data = None
+        if verbose:
+            timer.report_clear(f"extract convolved data for {det}")
+
+        del convolver
+
+        return convolved_data
+
+    def calibrate_signal(self, data, det, beam, convolved_data, verbose, detectors):
+        """By default, libConviqt results returns a signal that conforms to
+        TOD = (1 + epsilon) / 2 * intensity + (1 - epsilon) / 2 * polarization.
+
+        When calibrate = True, we rescale the TOD to
+        TOD = intensity + (1 - epsilon) / (1 + epsilon) * polarization
+        """
+        if not self.calibrate or beam.normalized():
+            return
+        timer = Timer()
+        timer.start()
+        offset = 0
+        for obs in data.obs:
+            if det not in obs.select_local_detectors(detectors):
+                continue
+            focalplane = obs.telescope.focalplane
+            epsilon = self._get_epsilon(focalplane, det)
+            # Make sure detector data output exists
+            obs.detdata.ensure(self.det_data, detectors=detectors)
+            # Loop over views
+            views = obs.view[self.view]
+            for view in views.detdata[self.det_data]:
+                nsample = len(view[det])
+                convolved_data[offset : offset + nsample] *= 2 / (1 + epsilon)
+                offset += nsample
+        if verbose:
+            timer.report_clear(f"calibrate detector {det}")
+        return
+
+    def cache(self, data, det, convolved_data, verbose, detectors):
+        """Inject the convolved data into the TOD cache."""
+        timer = Timer()
+        timer.start()
+        offset = 0
+        for obs in data.obs:
+            dets = obs.select_local_detectors(detectors)
+            if det not in dets:
+                continue
+            # Make sure detector data output exists
+            obs.detdata.ensure(self.det_data, detectors=detectors)
```

Comment:
Now this is not needed, since we do it once before the outer detector loop.

Reply:
Yes.